### PR TITLE
Route based links - master

### PR DESCRIPTION
**Pull request by lgebhardt**
Updates the links generation in LinkBuilder to use the routes url helper functions.

Adds `build_default_links` flags for resources and relationships to turn off building the default links.

When a route helper function can not be found the link is not built and a warning is emitted.

Updates the tests to work with this. This is the bulk of the changes.

Adds a `singleton` flag for singleton resources and a configurable way to resolve the singleton ids.

NaiveCache has been removed from the LinksBuilder. I wasn't seeing an appreciable performance difference with the way it was being used, and it may have been harming performance on typical requests where there is not duplication of the generated links. I do think we should be able to add a more long lived caching system that could improve performance.

There could also be a follow-on PR that adds an `except` flag to `jsonapi_relationships` along with a new `jsonapi_relationship` method to make it easier to create non-routable relationships.

### All Submissions:

- [ ] I've checked to ensure there aren't other open [Pull Requests](https://github.com/cerebris/jsonapi-resources/pulls) for the same update/change.
- [ ] I've submitted a [ticket](https://github.com/cerebris/jsonapi-resources/issues) for my issue if one did not already exist.
- [ ] My submission passes all tests. (Please run the full test suite locally to cut down on noise from travis failures.)
- [ ] I've used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message or the description.
- [ ] I've added/updated tests for this change.

### New Feature Submissions:

- [ ] I've submitted an issue that describes this feature, and received the go ahead from the maintainers.
- [ ] My submission includes new tests.
- [ ] My submission maintains compliance with [JSON:API](http://jsonapi.org/).

### Bug fixes and Changes to Core Features:

- [ ] I've included an explanation of what the changes do and why I'd like you to include them.
- [ ] I've provided test(s) that fails without the change.

### Test Plan:

Testing this on some real world apps should be done. It should be a drop in replacement. Assuming all resources are routable there should be no changes needed and no warnings emitted.

If there are some relationships or resources that do not have routes these will emit warnings when they are serialized (say from includes). The `build_default_links` option should be used:

```ruby
class EmployeeResource < JSONAPI::Resource
  build_default_links false

  has_many :expense_entries, build_default_links: false
end
```

Singleton resources now need to be declared as singletons with the `singleton` macro. This macro allows the singleton status (true or false) and options to be set. The only current option is a callable for the `singleton_key`.

```ruby
class PreferencesResource < JSONAPI::Resource
 ...

  singleton singleton_key: -> (context) {
    key = context[:current_user].try(:preferences).try(:id)
    raise JSONAPI::Exceptions::RecordNotFound.new(nil) if key.nil?
    key
  }

end
```
It's also possible to override the `singleton_key` method on the resource instead of providing a callable.

Since this PR is on master and most production apps are still running v0.9 you can test with #1247 as well. The core feature should be the same.

### Reviewer Checklist:
- [ ] Maintains compliance with JSON:API
- [ ] Adequate test coverage exists to prevent regressions

### CODE REVIEW

File: lib/jsonapi/basic_resource.rb [hunk 1]
```diff
@@ -629,6 +630,19 @@ def model_hint(model: _model_name, resource: _type)
         _model_hints[model.to_s.gsub('::', '/').underscore] = resource_type.to_s
       end
 
+      def singleton(*attrs)
+        @_singleton = (!!attrs[0] == attrs[0]) ? attrs[0] : true
```
Comment (dgeb) [inline lib/jsonapi/basic_resource.rb:634]: Is this effectively the same as `@_singleton = !!attrs[0]` ?

Comment (dgeb) [reply lib/jsonapi/basic_resource.rb:634]: Ah, perhaps you want it to be `true` if `attrs[0]` is `nil` - so just a bare `singleton` declaration

Comment (lgebhardt) [reply lib/jsonapi/basic_resource.rb:634]: I should ensure all combinations of this are tested.

**Review by dgeb — APPROVED** [review]
@lgebhardt looks good! I like the `:default` and `:none` options for `exclude_links`.

File: lib/jsonapi/relationship.rb [hunk 2]
```diff
@@ -114,7 +137,7 @@ def initialize(name, options = {})
 
       def to_s
         # :nocov: useful for debugging
-        "#{parent_resource._type}.#{name} => (#{belongs_to? ? 'ToOne' : 'BelongsToOne'}) #{resource_klass._type}"
+        "#{parent_resource}.#{name}(#{belongs_to? ? 'ToOne' : 'BelongsToOne'})"
```
Comment (dgeb) [inline lib/jsonapi/relationship.rb]:
```suggestion
        "#{parent_resource}.#{name}(#{belongs_to? ? 'BelongsToOne' : 'ToOne'})"
```

File: lib/jsonapi/resource_serializer.rb [hunk 3]
```diff
@@ -100,17 +100,12 @@ def serialize_to_links_hash(source, requested_relationship, resource_ids)
...
 
-      {
-          'links' => {
-              'self' => self_link(source, requested_relationship),
-              'related' => related_link(source, requested_relationship)
-          },
-          'data' => data
-      }
-    end
+      links_hash = { 'data': data }
```
Comment (dgeb) [inline lib/jsonapi/resource_serializer.rb]: should `links_hash` be `relationship_hash`?

Comment (lgebhardt) [reply lib/jsonapi/resource_serializer.rb]: Good catch. I thought I had fixed this naming issue a while ago.

**Pull request merged**